File: apps/api/internal/adapter/driven/mongodb/project_repository.go
package mongodb

import (
	"context"
	"time"

	"github.com/neurodynecorp/api/internal/domain/entity"
	"github.com/neurodynecorp/api/internal/domain/port"
	"github.com/neurodynecorp/api/pkg/metrics"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type ProjectRepository struct {
	collection *mongo.Collection
	metrics    *metrics.Metrics
}

func NewProjectRepository(client *Client, m *metrics.Metrics) *ProjectRepository {
	return &ProjectRepository{
		collection: client.Collection("projects"),
		metrics:    m,
	}
}

func (r *ProjectRepository) Create(ctx context.Context, project *entity.Project) error {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("projects", "insert").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("projects", "insert").Inc()
	}()

	_, err := r.collection.InsertOne(ctx, project)
	return err
}

func (r *ProjectRepository) GetByID(ctx context.Context, id string) (*entity.Project, error) {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("projects", "find_one").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("projects", "find_one").Inc()
	}()

	var project entity.Project
	err := r.collection.FindOne(ctx, bson.M{"_id": id}).Decode(&project)
	if err != nil {
		return nil, err
	}
	return &project, nil
}

func (r *ProjectRepository) Update(ctx context.Context, project *entity.Project) error {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("projects", "update").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("projects", "update").Inc()
	}()

	project.UpdatedAt = time.Now()
	_, err := r.collection.ReplaceOne(ctx, bson.M{"_id": project.ID}, project)
	return err
}

func (r *ProjectRepository) Delete(ctx context.Context, id string) error {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("projects", "delete").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("projects", "delete").Inc()
	}()

	_, err := r.collection.DeleteOne(ctx, bson.M{"_id": id})
	return err
}

func (r *ProjectRepository) ListByClient(ctx context.Context, clientID string, page, pageSize int) ([]*entity.Project, int64, error) {
	return r.list(ctx, bson.M{"client_id": clientID}, page, pageSize)
}

func (r *ProjectRepository) ListByStatus(ctx context.Context, status entity.ProjectStatus, page, pageSize int) ([]*entity.Project, int64, error) {
	return r.list(ctx, bson.M{"status": status}, page, pageSize)
}

func (r *ProjectRepository) List(ctx context.Context, filter port.ProjectFilter) ([]*entity.Project, int64, error) {
	query := bson.M{}
	if filter.ClientID != "" {
		query["client_id"] = filter.ClientID
	}
	if filter.Status != "" {
		query["status"] = filter.Status
	}
	if filter.Type != "" {
		query["type"] = filter.Type
	}
	if filter.Search != "" {
		query["$or"] = []bson.M{
			{"title": bson.M{"$regex": filter.Search, "$options": "i"}},
			{"description": bson.M{"$regex": filter.Search, "$options": "i"}},
		}
	}
	return r.list(ctx, query, filter.Page, filter.PageSize)
}

func (r *ProjectRepository) list(ctx context.Context, query bson.M, page, pageSize int) ([]*entity.Project, int64, error) {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("projects", "find").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("projects", "find").Inc()
	}()

	total, err := r.collection.CountDocuments(ctx, query)
	if err != nil {
		return nil, 0, err
	}

	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 20
	}

	opts := options.Find().
		SetSkip(int64((page - 1) * pageSize)).
		SetLimit(int64(pageSize)).
		SetSort(bson.M{"created_at": -1})

	cursor, err := r.collection.Find(ctx, query, opts)
	if err != nil {
		return nil, 0, err
	}
	defer cursor.Close(ctx)

	var projects []*entity.Project
	if err := cursor.All(ctx, &projects); err != nil {
		return nil, 0, err
	}

	return projects, total, nil
}

var _ port.ProjectRepository = (*ProjectRepository)(nil)

File: apps/server/src/adapter/driven/mongodb/project-repository.ts
import { ObjectId } from "mongodb";
import type { Collection, Filter } from "mongodb";
import type { Project } from "../../../domain/entity/index.js";
import type { ProjectRepository, ProjectFilter } from "../../../domain/port/index.js";
import type { MongoDBClient } from "./client.js";

interface ProjectDoc {
  _id: ObjectId;
  name: string;
  description: string;
  client_id: string;
  status: string;
  type: string;
  features: Project["features"];
  user_roles: string[];
  integrations: string[];
  budget_range: Project["budgetRange"];
  timeline: Project["timeline"];
  attachments: Project["attachments"];
  assigned_team: string[];
  specification_id?: string;
  progress: number;
  milestones: Project["milestones"];
  created_at: Date;
  updated_at: Date;
}

function toDoc(p: Project): ProjectDoc {
  return {
    _id: new ObjectId(p.id),
    // ProjectService names the field `title`; storage uses `name`.
    name: p.name ?? (p as unknown as { title?: string }).title ?? "",
    description: p.description,
    client_id: p.clientId,
    status: p.status,
    type: p.type,
    features: p.features,
    user_roles: p.userRoles,
    integrations: p.integrations,
    budget_range: p.budgetRange,
    timeline: p.timeline,
    attachments: p.attachments,
    assigned_team: p.assignedTeam,
    specification_id: p.specificationId,
    progress: p.progress,
    milestones: p.milestones,
    created_at: p.createdAt,
    updated_at: p.updatedAt,
  };
}

function fromDoc(d: ProjectDoc): Project {
  return {
    id: d._id.toHexString(),
    name: d.name,
    description: d.description,
    clientId: d.client_id,
    status: d.status as Project["status"],
    type: d.type as Project["type"],
    features: d.features,
    userRoles: d.user_roles,
    integrations: d.integrations,
    budgetRange: d.budget_range,
    timeline: d.timeline,
    attachments: d.attachments,
    assignedTeam: d.assigned_team,
    specificationId: d.specification_id,
    progress: d.progress,
    milestones: d.milestones,
    createdAt: d.created_at,
    updatedAt: d.updated_at,
  };
}

export class MongoProjectRepository implements ProjectRepository {
  private readonly col: Collection<ProjectDoc>;

  constructor(client: MongoDBClient) {
    this.col = client.collection<ProjectDoc>("projects");
  }

  async findById(id: string): Promise<Project | null> {
    const doc = await this.col.findOne({ _id: new ObjectId(id) });
    return doc ? fromDoc(doc) : null;
  }

  async findAll(filter?: ProjectFilter): Promise<Project[]> {
    const query: Filter<ProjectDoc> = {};

    if (filter?.clientId) {
      query.client_id = filter.clientId;
    }
    if (filter?.status) {
      query.status = filter.status;
    }
    if (filter?.type) {
      query.type = filter.type;
    }
    if (filter?.assignedTeamMember) {
      query.assigned_team = filter.assignedTeamMember;
    }
    if (filter?.search) {
      query.$or = [
        { name: { $regex: filter.search, $options: "i" } },
        { description: { $regex: filter.search, $options: "i" } },
      ];
    }

    const docs = await this.col
      .find(query)
      .sort({ created_at: -1 })
      .toArray();

    return docs.map(fromDoc);
  }

  async findByClientId(clientId: string): Promise<Project[]> {
    const docs = await this.col
      .find({ client_id: clientId })
      .sort({ created_at: -1 })
      .toArray();

    return docs.map(fromDoc);
  }

  async create(project: Project): Promise<Project> {
    await this.col.insertOne(toDoc(project));
    return project;
  }

  async update(idOrProject: string | Project, data?: Partial<Project>): Promise<Project> {
    if (typeof idOrProject === "string") {
      const existing = await this.findById(idOrProject);
      if (!existing) throw new Error(`Project not found: ${idOrProject}`);
      const merged = { ...existing, ...data, updatedAt: new Date() };
      const doc = toDoc(merged);
      await this.col.replaceOne({ _id: doc._id }, doc);
      return merged;
    }
    const doc = toDoc({ ...idOrProject, updatedAt: new Date() });
    await this.col.replaceOne({ _id: doc._id }, doc);
    return { ...idOrProject, updatedAt: doc.updated_at };
  }

  async list(filter: ProjectFilter & { page?: number; pageSize?: number }): Promise<{ projects: Project[]; total: number }> {
    const query: Filter<ProjectDoc> = {};
    if (filter?.clientId) query.client_id = filter.clientId;
    if (filter?.status) query.status = filter.status;
    if (filter?.type) query.type = filter.type;
    if (filter?.assignedTeamMember) query.assigned_team = filter.assignedTeamMember;
    if (filter?.search) {
      query.$or = [
        { name: { $regex: filter.search, $options: "i" } },
        { description: { $regex: filter.search, $options: "i" } },
      ];
    }

    const page = filter?.page ?? 1;
    const pageSize = filter?.pageSize ?? 20;
    const total = await this.col.countDocuments(query);
    const docs = await this.col
      .find(query)
      .sort({ created_at: -1 })
      .skip((page - 1) * pageSize)
      .limit(pageSize)
      .toArray();

    return { projects: docs.map(fromDoc), total };
  }

  async delete(id: string): Promise<void> {
    await this.col.deleteOne({ _id: new ObjectId(id) });
  }
}
